src/input/isf.c

```diff
@@ -0,0 +1,796 @@
+ */
+
+#include <config.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <ctype.h>
+#include <string.h>
+#include <stdint.h>
+#include <math.h>
+#include <errno.h>
+#include <limits.h>
+#include <libsigrok/libsigrok.h>
+#include "libsigrok-internal.h"
+
+#define LOG_PREFIX "input/isf"
+
+#define CHUNK_SIZE	(4 * 1024 * 1024)
+
+/* Maximum header size. */
+#define MAX_HEADER_SIZE 1024
+
+/* Number of items in the header. */
+#define HEADER_ITEMS_PARAMETERS 10
+
+/* Maximum length of a channel name. */
+#define MAX_CHANNEL_NAME_SIZE 32
+
+/* Maximum size of encoding and waveform strings. */
+#define MAX_ENCODING_STRING_SIZE 10
+#define MAX_WAVEFORM_STRING_SIZE 10
+
+/* Maximum number of bytes per sample. */
+#define MAX_INT_BYTNR 8
+#define FLOAT_BYTNR 4
+
+/* Size of buffer in which byte order and data format strings are stored. */
+#define BYTE_ORDER_BUFFER_SIZE 4
+#define DATA_FORMAT_BUFFER_SIZE 3
+
+/* Byte order */
+enum byteorder {
+	LSB = 0,
+	MSB = 1,
+};
+
+/* Format, i.e. RI (signed integer), RP (unsigned integer) or FP (floating point) */
+enum format {
+	RI = 0,
+	RP = 1,
+	FP = 2,
+};
+
+/* Waveform type, i.e. analog or radar frequency (RF) */
+enum waveform_type {
+	ANALOG = 0,
+	RF_FD = 1,
+};
+
+union floating_point {
+	float f;
+	uint32_t i;
+};
+
+struct context {
+	gboolean started;
+	gboolean create_channel;
+	gboolean found_data_section;
+	float yoff;
+	float yzero;
+	float ymult;
+	float xincr;
+	unsigned int bytnr;
+	enum byteorder byte_order;
+	enum format bn_fmt;
+	enum waveform_type wfmtype;
+	char channel_name[MAX_CHANNEL_NAME_SIZE];
+};
+
+/*
+ * Header items used to process the input file.
+ *
+ * Parameter WFID is optional, the rest are required.
+ */
+enum header_items_enum {
+	YOFF = 0,
+	YZERO = 1,
+	YMULT = 2,
+	XINCR = 3,
+	BYTNR = 4,
+	BYTE_ORDER = 5,
+	BN_FMT = 6,
+	WFID = 7,
+	WFMTYPE = 8,
+	ENCODING = 9,
+};
+
+/* Strings searched for in the file header representing the header items. */
+static const char *header_items[] = {
+	[YOFF] = "YOFF ",
+	[YZERO] = "YZERO ",
+	[YMULT] = "YMULT ",
+	[XINCR] = "XINCR ",
+	[BYTNR] = "BYT_NR ",
+	[BYTE_ORDER] = "BYT_OR ",
+	[BN_FMT] = "BN_FMT ",
+	[WFID] = "WFID ",
+	[WFMTYPE] = "WFMTYPE ",
+	[ENCODING] = "ENCDG ",
+};
+
+/* Find the header item string in the header. */
+static char *find_item(const char *buf, size_t buflen, const char *item)
+{
+	return g_strstr_len(buf, buflen, item);
+}
+
+/* Find curve which marks the end of the header and the start of the data. */
+static char *find_data_section(GString *buf)
+{
+	const char curve[] = "CURVE #";
+
+	char *data_ptr;
+	size_t offset, metadata_length;
+
+	data_ptr = g_strstr_len(buf->str, buf->len, curve);
+	if (data_ptr == NULL)
+		return NULL;
+
+	data_ptr += strlen(curve);
+	offset = data_ptr - buf->str;
+	if (offset >= buf->len)
+		return NULL;
+
+	/* Curve metadata length is an ASCII byte, hence -48. */
+	metadata_length = (size_t) *data_ptr - 48;
```

Comment:
do you mean the length is encoded as an ascii digit '0' to '9' ? or that they simply mapped lengths 0 to (255-48) to chars 0x30 to 0xff ?

Suggest bounds checking before subtracting, in case of malformed packet.

Reply:
The length is encoded as an ascii digit '0' to '9'. Fixed the comment and bounds checking is performed.

Reply:
Instead of using `48` you should use `'0'` to make this obvious.

Reply:
> Instead of using `48` you should use `'0'` to make this obvious.

Agreed. This is something often written `if (val > '9' ||  val < '0') return bad;  metadata_length = val - '0'`

Reply:
Replaced 48 with '0' to make it obvious.